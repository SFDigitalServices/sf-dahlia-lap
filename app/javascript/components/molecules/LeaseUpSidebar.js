import React, { useState } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import LeaseUpSidebarButtons from './LeaseUpSidebarButtons'
import StatusItems from './StatusItems'

import StatusItemShape from '../../utils/shapes/StatusItemShape'

const MAX_UPDATES_TO_SHOW_DEFAULT = 4

const getMostRecentStatus = (statusHistory) =>
  statusHistory && statusHistory[0]
    ? statusHistory[0].status
    : null

// visible for testing
export const ShowOrHideStatusesButton = ({ showingAllStatuses, onClick }) => {
  const buttonClasses = classNames(
    'button-link',
    't-tiny',
    'show-all-updates-toggle'
  )

  return (
    <button
      className={buttonClasses}
      type='button'
      onClick={onClick}
    >
      {showingAllStatuses ? 'Show only recent status updates' : 'Show all status updates'}
    </button>
  )
}

const LeaseUpSidebar = ({ isLoading, statusItems, onSaveClicked, onChangeStatus, onAddCommentClicked }) => {
  const [showingAllStatuses, setShowingAllStatuses] = useState(false)

  const currentStatus = getMostRecentStatus(statusItems)

  const sidebarButtons = (withMobileStyling) => (
    <div className={withMobileStyling ? 'hide-large-up' : 'show-large-up'}>
      <LeaseUpSidebarButtons
        isLoading={isLoading}
        status={currentStatus}
        withMobileStyling={withMobileStyling}
        onSaveClicked={onSaveClicked}
        onChangeStatus={onChangeStatus}
        onAddCommentClicked={onAddCommentClicked}
      />
    </div>
  )
  const onShowHideStatusesToggled = () => setShowingAllStatuses(!showingAllStatuses)

  const numberOfStatusesToDisplay = showingAllStatuses ? statusItems.length : MAX_UPDATES_TO_SHOW_DEFAULT

  return (
    <div className='sidebar-content'>
      {sidebarButtons(false)}
      <h3 className='status-history-label'>Status History</h3>
      <StatusItems
        statusItems={statusItems}
        limit={numberOfStatusesToDisplay}
      />
      {statusItems.length > MAX_UPDATES_TO_SHOW_DEFAULT && (
        <ShowOrHideStatusesButton
          showingAllStatuses={showingAllStatuses}
          onClick={onShowHideStatusesToggled}
        />
      )}
      {sidebarButtons(true)}
    </div>
  )
}

LeaseUpSidebar.propTypes = {
  isLoading: PropTypes.bool,
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape)),
  onSaveClicked: PropTypes.func,
  onChangeStatus: PropTypes.func,
  onAddCommentClicked: PropTypes.func
}

LeaseUpSidebar.defaultProps = {
  isLoading: false,
  statusItems: [],
  onSaveClicked: null,
  onChangeStatus: null,
  onAddCommentClicked: null
}

export default LeaseUpSidebar
