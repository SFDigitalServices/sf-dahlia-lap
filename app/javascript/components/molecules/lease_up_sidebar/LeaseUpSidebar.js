import React from 'react'

import { isEmpty } from 'lodash'
import PropTypes from 'prop-types'

import StatusItemShape from '../../../utils/shapes/StatusItemShape'
import LeaseUpStatusButtons from './LeaseUpStatusButtons'
import StatusHistoryContainer from './StatusHistoryContainer'

const getMostRecentStatus = (statusHistory) =>
  statusHistory && statusHistory[0] ? statusHistory[0].status : null

const LeaseUpSidebar = ({
  isLoading,
  statusItems,
  onSaveClicked,
  onChangeStatus,
  onAddCommentClicked
}) => {
  const currentStatus = getMostRecentStatus(statusItems)

  const saveButton = (
    <button
      className='primary save-button'
      id='save-supplemental-application'
      type='submit'
      disabled={isLoading}
      onClick={onSaveClicked}
    >
      {isLoading ? 'Savingâ€¦' : 'Save'}
    </button>
  )

  const statusButtons = (
    <LeaseUpStatusButtons
      isLoading={isLoading}
      status={currentStatus}
      onChangeStatus={onChangeStatus}
      onAddCommentClicked={onAddCommentClicked}
    />
  )

  return (
    <div className='sidebar-content'>
      <div className={'show-large-up'}>
        {saveButton}
        {statusButtons}
      </div>
      {!isEmpty(statusItems) && <StatusHistoryContainer statusItems={statusItems} />}
      <div className={'hide-large-up'}>
        {statusButtons}
        {saveButton}
      </div>
    </div>
  )
}

LeaseUpSidebar.propTypes = {
  isLoading: PropTypes.bool,
  statusItems: PropTypes.arrayOf(PropTypes.shape(StatusItemShape)),
  onSaveClicked: PropTypes.func,
  onChangeStatus: PropTypes.func,
  onAddCommentClicked: PropTypes.func
}

LeaseUpSidebar.defaultProps = {
  isLoading: false,
  statusItems: [],
  onSaveClicked: null,
  onChangeStatus: null,
  onAddCommentClicked: null
}

export default LeaseUpSidebar
