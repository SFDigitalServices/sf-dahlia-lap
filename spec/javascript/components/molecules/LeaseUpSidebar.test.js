import React from 'react'
import renderer from 'react-test-renderer'
import LeaseUpSidebar from 'components/molecules/LeaseUpSidebar'
import { mockStatusItem, mockStatusItems, mockManyStatusItems } from '../../mocks/statusItemMock'

const getWrapper = (currentStatus, items) => renderer.create(
  <LeaseUpSidebar
    currentStatus={currentStatus}
    statusItems={items}
  />
)

describe('LeaseUpSidebar', () => {
  test('should render with empty status items correctly', () => {
    const wrapper = getWrapper('Approved', [])
    expect(wrapper).toMatchSnapshot()
  })

  test('should render with a single status item correctly', () => {
    const wrapper = getWrapper('Approved', [mockStatusItem()])
    expect(wrapper).toMatchSnapshot()
  })

  test('should render with multiple status items correctly', () => {
    const wrapper = getWrapper('Approved', mockStatusItems())
    expect(wrapper).toMatchSnapshot()
  })

  test('should render with more than 4 status items correctly', () => {
    const wrapper = getWrapper('Approved', mockManyStatusItems(5))
    expect(wrapper).toMatchSnapshot()
  })
})
