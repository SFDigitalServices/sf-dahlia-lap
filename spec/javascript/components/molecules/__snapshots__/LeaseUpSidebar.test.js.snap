// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaseUpSidebar should render with a single status item correctly 1`] = `
<div
  className="sidebar-content"
>
  <div
    className="show-large-up"
  >
    <button
      className="primary save-button"
      onClick={null}
    >
      Save
    </button>
    <div
      className="status-history-buttons"
    >
      <div
        className="dropdown"
        onClick={[Function]}
      >
        <button
          aria-expanded="false"
          className="button dropdown-button has-icon--right text-align-left is-approved tight-padding"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ui-icon ui-small"
          >
            <svg>
              <use
                xlinkHref="#i-arrow-down"
              />
            </svg>
          </span>
          Approved
        </button>
        <div
          aria-hidden="true"
          className="dropdown-menu-wrapper"
          role="menu"
        />
      </div>
      <button
        className="tertiary tight-padding"
        onClick={null}
      >
        Add a comment
      </button>
    </div>
  </div>
  <h3
    className="status-history-label"
  >
    Status History
  </h3>
  <div
    className="status-items"
  >
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-approved"
          >
            Approved
          </div>
          <div
            className="status-item-date"
          >
            August 25th 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Approval letter sent
        </div>
        <div
          className="status-item-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
        </div>
      </div>
    </div>
  </div>
  <div
    className="hide-large-up"
  >
    <div
      className="status-history-buttons"
    >
      <div
        className="dropdown"
        onClick={[Function]}
      >
        <button
          aria-expanded="false"
          className="button dropdown-button has-icon--right text-align-left is-approved tight-padding"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ui-icon ui-small"
          >
            <svg>
              <use
                xlinkHref="#i-arrow-down"
              />
            </svg>
          </span>
          Approved
        </button>
        <div
          aria-hidden="true"
          className="dropdown-menu-wrapper"
          role="menu"
        />
      </div>
      <button
        className="tertiary tight-padding"
        onClick={null}
      >
        Add a comment
      </button>
    </div>
    <button
      className="primary save-button"
      onClick={null}
    >
      Save
    </button>
  </div>
</div>
`;

exports[`LeaseUpSidebar should render with empty status items correctly 1`] = `
<div
  className="sidebar-content"
>
  <div
    className="show-large-up"
  >
    <button
      className="primary save-button"
      onClick={null}
    >
      Save
    </button>
    <div
      className="status-history-buttons"
    >
      <div
        className="dropdown"
        onClick={[Function]}
      >
        <button
          aria-expanded="false"
          className="button dropdown-button has-icon--right text-align-left is-approved tight-padding"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ui-icon ui-small"
          >
            <svg>
              <use
                xlinkHref="#i-arrow-down"
              />
            </svg>
          </span>
          Approved
        </button>
        <div
          aria-hidden="true"
          className="dropdown-menu-wrapper"
          role="menu"
        />
      </div>
      <button
        className="tertiary tight-padding"
        onClick={null}
      >
        Add a comment
      </button>
    </div>
  </div>
  <h3
    className="status-history-label"
  >
    Status History
  </h3>
  <div
    className="status-items"
  />
  <div
    className="hide-large-up"
  >
    <div
      className="status-history-buttons"
    >
      <div
        className="dropdown"
        onClick={[Function]}
      >
        <button
          aria-expanded="false"
          className="button dropdown-button has-icon--right text-align-left is-approved tight-padding"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ui-icon ui-small"
          >
            <svg>
              <use
                xlinkHref="#i-arrow-down"
              />
            </svg>
          </span>
          Approved
        </button>
        <div
          aria-hidden="true"
          className="dropdown-menu-wrapper"
          role="menu"
        />
      </div>
      <button
        className="tertiary tight-padding"
        onClick={null}
      >
        Add a comment
      </button>
    </div>
    <button
      className="primary save-button"
      onClick={null}
    >
      Save
    </button>
  </div>
</div>
`;

exports[`LeaseUpSidebar should render with more than 4 status items correctly 1`] = `
<div
  className="sidebar-content"
>
  <div
    className="show-large-up"
  >
    <button
      className="primary save-button"
      onClick={null}
    >
      Save
    </button>
    <div
      className="status-history-buttons"
    >
      <div
        className="dropdown"
        onClick={[Function]}
      >
        <button
          aria-expanded="false"
          className="button dropdown-button has-icon--right text-align-left is-approved tight-padding"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ui-icon ui-small"
          >
            <svg>
              <use
                xlinkHref="#i-arrow-down"
              />
            </svg>
          </span>
          Approved
        </button>
        <div
          aria-hidden="true"
          className="dropdown-menu-wrapper"
          role="menu"
        />
      </div>
      <button
        className="tertiary tight-padding"
        onClick={null}
      >
        Add a comment
      </button>
    </div>
  </div>
  <h3
    className="status-history-label"
  >
    Status History
  </h3>
  <div
    className="status-items"
  >
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-approved"
          >
            Approved
          </div>
          <div
            className="status-item-date"
          >
            August 25th 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Approval letter sent
        </div>
        <div
          className="status-item-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
        </div>
      </div>
    </div>
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-approved"
          >
            Approved
          </div>
          <div
            className="status-item-date"
          >
            August 25th 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Approval letter sent
        </div>
        <div
          className="status-item-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
        </div>
      </div>
    </div>
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-approved"
          >
            Approved
          </div>
          <div
            className="status-item-date"
          >
            August 25th 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Approval letter sent
        </div>
        <div
          className="status-item-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
        </div>
      </div>
    </div>
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-approved"
          >
            Approved
          </div>
          <div
            className="status-item-date"
          >
            August 25th 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Approval letter sent
        </div>
        <div
          className="status-item-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
        </div>
      </div>
    </div>
  </div>
  <button
    className="button-link t-tiny show-all-updates-toggle"
    onClick={[Function]}
  >
    Show all status updates
  </button>
  <div
    className="hide-large-up"
  >
    <div
      className="status-history-buttons"
    >
      <div
        className="dropdown"
        onClick={[Function]}
      >
        <button
          aria-expanded="false"
          className="button dropdown-button has-icon--right text-align-left is-approved tight-padding"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ui-icon ui-small"
          >
            <svg>
              <use
                xlinkHref="#i-arrow-down"
              />
            </svg>
          </span>
          Approved
        </button>
        <div
          aria-hidden="true"
          className="dropdown-menu-wrapper"
          role="menu"
        />
      </div>
      <button
        className="tertiary tight-padding"
        onClick={null}
      >
        Add a comment
      </button>
    </div>
    <button
      className="primary save-button"
      onClick={null}
    >
      Save
    </button>
  </div>
</div>
`;

exports[`LeaseUpSidebar should render with multiple status items correctly 1`] = `
<div
  className="sidebar-content"
>
  <div
    className="show-large-up"
  >
    <button
      className="primary save-button"
      onClick={null}
    >
      Save
    </button>
    <div
      className="status-history-buttons"
    >
      <div
        className="dropdown"
        onClick={[Function]}
      >
        <button
          aria-expanded="false"
          className="button dropdown-button has-icon--right text-align-left is-approved tight-padding"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ui-icon ui-small"
          >
            <svg>
              <use
                xlinkHref="#i-arrow-down"
              />
            </svg>
          </span>
          Approved
        </button>
        <div
          aria-hidden="true"
          className="dropdown-menu-wrapper"
          role="menu"
        />
      </div>
      <button
        className="tertiary tight-padding"
        onClick={null}
      >
        Add a comment
      </button>
    </div>
  </div>
  <h3
    className="status-history-label"
  >
    Status History
  </h3>
  <div
    className="status-items"
  >
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-approved"
          >
            Approved
          </div>
          <div
            className="status-item-date"
          >
            August 25th 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Approval letter sent
        </div>
        <div
          className="status-item-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
        </div>
      </div>
    </div>
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-approved"
          >
            Approved
          </div>
          <div
            className="status-item-date"
          >
            August 24th 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Unit selected
        </div>
        <div
          className="status-item-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
        </div>
      </div>
    </div>
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-approved"
          >
            Approved
          </div>
          <div
            className="status-item-date"
          >
            August 23rd 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Waiting for subsidy inspection
        </div>
        <div
          className="status-item-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque bibendum condimentum lorem consectetur eleifend.
        </div>
      </div>
    </div>
    <div
      className="status-items"
    >
      <div
        className="status-item"
      >
        <div
          className="status-item-header"
        >
          <div
            className="status-pill is-waitlisted"
          >
            Waitlisted
          </div>
          <div
            className="status-item-date"
          >
            August 22nd 2020
          </div>
        </div>
        <div
          className="item-substatus-text"
        >
          Written confirmation sent
        </div>
        <div
          className="status-item-text"
        />
      </div>
    </div>
  </div>
  <button
    className="button-link t-tiny show-all-updates-toggle"
    onClick={[Function]}
  >
    Show all status updates
  </button>
  <div
    className="hide-large-up"
  >
    <div
      className="status-history-buttons"
    >
      <div
        className="dropdown"
        onClick={[Function]}
      >
        <button
          aria-expanded="false"
          className="button dropdown-button has-icon--right text-align-left is-approved tight-padding"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ui-icon ui-small"
          >
            <svg>
              <use
                xlinkHref="#i-arrow-down"
              />
            </svg>
          </span>
          Approved
        </button>
        <div
          aria-hidden="true"
          className="dropdown-menu-wrapper"
          role="menu"
        />
      </div>
      <button
        className="tertiary tight-padding"
        onClick={null}
      >
        Add a comment
      </button>
    </div>
    <button
      className="primary save-button"
      onClick={null}
    >
      Save
    </button>
  </div>
</div>
`;
